Python/bytecodes.c

```diff
@@ -3347,22 +3353,22 @@ dummy_func(
             }
         }
 
-        op(_CALL_BUILTIN_CLASS, (callable, self_or_null, args[oparg] -- res)) {
+        op(_CALL_BUILTIN_CLASS, (callable: _Py_TaggedObject, self_or_null, args[oparg] -- res)) {
             int total_args = oparg;
             if (self_or_null != NULL) {
                 args--;
                 total_args++;
             }
-            DEOPT_IF(!PyType_Check(callable));
-            PyTypeObject *tp = (PyTypeObject *)callable;
+            DEOPT_IF(!PyType_Check(Py_CLEAR_TAG(callable)));
+            PyTypeObject *tp = (PyTypeObject *)Py_CLEAR_TAG(callable);
             DEOPT_IF(tp->tp_vectorcall == NULL);
             STAT_INC(CALL, hit);
-            res = tp->tp_vectorcall((PyObject *)tp, args, total_args, NULL);
+            res = PyObject_TypeVectorcall_Tagged(tp, args, total_args, NULL);
```

Comment:
I think instead of adding a new `PyObject_TypeVectorcall_Tagged` that's only called here, we should inline the `#ifdef` and just use `PyObject_Vectorcall_Tagged` in the free-threaded build. For example:

#if defined(Py_GIL_DISABLED) || defined(Py_OBJECT_TEST_TAG)
    res = PyObject_Vectorcall_Tagged((PyObject *)tp, args, total_args, NULL);
#else 
    res = tp->tp_vectorcall((PyObject *)tp, (PyObject **)args, total_args, NULL);
#endif

This might be marginally slower in the free-threaded build, but should be faster in the default build by avoiding a non-inlinable call.

Reply:
With LTO it should be inlined though?